### PR TITLE
Fix set avatar from gallery

### PR DESCRIPTION
**Pull request by mr-ant**
Fixes avatar gallery,when use storage(like AWS S3) .

### CODE REVIEW

File: misago/users/avatars/gallery.py [hunk 1]
```diff
@@ -66,7 +68,12 @@ def load_avatar_galleries():
 
 
 def set_avatar(user, image):
-    store.store_new_avatar(user, Image.open(image.path))
+    try:
+        response = requests.get(image.url)
```
Comment (rafalp) [inline misago/users/avatars/gallery.py]: `requests` should be unnecessary here - value of `File` and `Image` model fields is `File` instance that has `open()` method that delegates opening/reading operation to your configured `STATICFILES_STORAGE`:

```
with image.open() as img:
    store.store_new_avatar(user, Image.open(img))
```

Could you please say if that code works? I don't have S3 to test it out myself.

**Comment by mr-ant** [comment]
I have tested OSS, and QiNiu.S3 is too slow to visit in China. I will try it tomorrow.

**Comment by mr-ant** [comment]
I feel that your code is invalid, because when the image is uploaded directly to the cloud storage, it will not go to the local, it should not be used to open. My current approach is to load the file into memory via an image link and then open the image from memory. I will test it together tomorrow. Thanks Reply.

**Comment by rafalp** [comment]
@mr-ant `File` object that Django uses is only a proxy to file stored in storage. If storage is S3 or other cloud storage, `.open()` will download it from 3rd party service, so you don't have to do it yourself.

**Comment by mr-ant** [comment]
Hi @rafalp 
`with image.open() as img:
    store.store_new_avatar(user, Image.open(img))`
This code is not work,because 'AvatarGallery' object has no attribute 'open'.

**Comment by mr-ant** [comment]
Hi @rafalp :

**Comment by mr-ant** [comment]
Hi @rafalp
test with S3,and worked!

**Comment by rafalp** [comment]
@mr-ant 

> This code is not work,because 'AvatarGallery' object has no attribute 'open'.

Yeah, `AvatarGallery` object pretends its file, but not completely. Could you please also test:

```
with image.image.open() as img:
    store.store_new_avatar(user, Image.open(img))
```

I really want to avoid maintaining file download mechanic on our side, if there's tested one in Django.

**Comment by mr-ant** [comment]
@rafalp
Tested and get AttributeError: __enter__,but thanks for notice ,Find a way to make this work by modify your code.
```
with image.image.file as img:
    store.store_new_avatar(user, Image.open(img))
```

**Comment by rafalp** [comment]
@mr-ant plase take a note that it should be `image.image.open()`

**Comment by mr-ant** [comment]
@rafalp
`image.image.open()` will caise an error `AttributeError: __enter__`
```
def set_avatar(user, image):
    img = image.image.file
    store.store_new_avatar(user, Image.open(img))
```
if use storage, `img` argument will  be a storage file object, else will be a file object.so anyway those code above will work.

**Comment by rafalp** [comment]
@mr-ant The actual code should be:

```
def set_avatar(user, image):
    with image.image.open() as img:
        store.store_new_avatar(user, Image.open(img))
```

**Comment by mr-ant** [comment]
@rafalp
`with image.image.open() as img:` will caise an error called `AttributeError: __enter__` when using storage or not.
here is trackback blow:
File "/usr/local/lib/python3.6/site-packages/rest_framework/viewsets.py", line 95, in view
django_1    |     return self.dispatch(request, *args, **kwargs)
django_1    |   File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 494, in dispatch
django_1    |     response = self.handle_exception(exc)
django_1    |   File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 454, in handle_exception
django_1    |     self.raise_uncaught_exception(exc)
django_1    |   File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 491, in dispatch
django_1    |     response = handler(request, *args, **kwargs)
django_1    |   File "/app/misago/users/api/users.py", line 102, in avatar
django_1    |     return avatar_endpoint(request)
django_1    |   File "/app/misago/core/decorators.py", line 30, in decorator
django_1    |     return f(request, *args, **kwargs)
django_1    |   File "/app/misago/users/api/userendpoints/avatar.py", line 35, in avatar_endpoint
django_1    |     return avatar_post(avatar_options, request.user, request.data)
django_1    |   File "/app/misago/users/api/userendpoints/avatar.py", line 126, in avatar_post
django_1    |     response_dict = {'detail': rpc_handler(user, data)}
django_1    |   File "/app/misago/users/api/userendpoints/avatar.py", line 162, in avatar_gallery
django_1    |     avatars.gallery.set_avatar(user, image)
django_1    |   File "/app/misago/users/avatars/gallery.py", line 68, in set_avatar
django_1    |     with image.image.open() as img:
django_1    | AttributeError: __enter__

**Comment by rafalp** [comment]
Okay, I see now that `django.core.files.base.File` doesn't work as context manager. But it still provides file-pointer functions like open, close, read, seek. So this code should be correct:

```
def set_avatar(user, image):
    fp =image.image.open()
    store.store_new_avatar(user, Image.open(fp))
    fp.close()
```

**Comment by mr-ant** [comment]
hi @rafalp
sorry,is not correct, here is trackback below:
n
django_1    |     response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
django_1    |   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 139, in handle_uncaught_exception
django_1    |     return debug.technical_500_response(request, *exc_info)
django_1    |   File "/usr/local/lib/python3.6/site-packages/django_extensions/management/technical_response.py", line 6, in null_technical_500_response
django_1    |     six.reraise(exc_type, exc_value, tb)
django_1    |   File "/usr/local/lib/python3.6/site-packages/six.py", line 693, in reraise
django_1    |     raise value
django_1    |   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 41, in inner
django_1    |     response = get_response(request)
django_1    |   File "/usr/local/lib/python3.6/site-packages/django/utils/deprecation.py", line 140, in __call__
django_1    |     response = self.get_response(request)
django_1    |   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 43, in inner
django_1    |     response = response_for_exception(request, exc)
django_1    |   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 93, in response_for_exception
django_1    |     response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
django_1    |   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 139, in handle_uncaught_exception
django_1    |     return debug.technical_500_response(request, *exc_info)
django_1    |   File "/usr/local/lib/python3.6/site-packages/django_extensions/management/technical_response.py", line 6, in null_technical_500_response
django_1    |     six.reraise(exc_type, exc_value, tb)
django_1    |   File "/usr/local/lib/python3.6/site-packages/six.py", line 693, in reraise
django_1    |     raise value
django_1    |   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 41, in inner
django_1    |     response = get_response(request)
django_1    |   File "/usr/local/lib/python3.6/site-packages/django/utils/deprecation.py", line 140, in __call__
django_1    |     response = self.get_response(request)
django_1    |   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 43, in inner
django_1    |     response = response_for_exception(request, exc)
django_1    |   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 93, in response_for_exception
django_1    |     response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
django_1    |   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 139, in handle_uncaught_exception
django_1    |     return debug.technical_500_response(request, *exc_info)
django_1    |   File "/usr/local/lib/python3.6/site-packages/django_extensions/management/technical_response.py", line 6, in null_technical_500_response
django_1    |     six.reraise(exc_type, exc_value, tb)
django_1    |   File "/usr/local/lib/python3.6/site-packages/six.py", line 693, in reraise
django_1    |     raise value
django_1    |   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 41, in inner
django_1    |     response = get_response(request)
django_1    |   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py", line 187, in _get_response
django_1    |     response = self.process_exception_by_middleware(e, request)
django_1    |   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py", line 185, in _get_response
django_1    |     response = wrapped_callback(request, *callback_args, **callback_kwargs)
django_1    |   File "/usr/local/lib/python3.6/contextlib.py", line 52, in inner
django_1    |     return func(*args, **kwds)
django_1    |   File "/usr/local/lib/python3.6/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
django_1    |     return view_func(*args, **kwargs)
django_1    |   File "/usr/local/lib/python3.6/site-packages/rest_framework/viewsets.py", line 95, in view
django_1    |     return self.dispatch(request, *args, **kwargs)
django_1    |   File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 494, in dispatch
django_1    |     response = self.handle_exception(exc)
django_1    |   File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 454, in handle_exception
django_1    |     self.raise_uncaught_exception(exc)
django_1    |   File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 491, in dispatch
django_1    |     response = handler(request, *args, **kwargs)
django_1    |   File "/app/misago/users/api/users.py", line 102, in avatar
django_1    |     return avatar_endpoint(request)
django_1    |   File "/app/misago/core/decorators.py", line 30, in decorator
django_1    |     return f(request, *args, **kwargs)
django_1    |   File "/app/misago/users/api/userendpoints/avatar.py", line 35, in avatar_endpoint
django_1    |     return avatar_post(avatar_options, request.user, request.data)
django_1    |   File "/app/misago/users/api/userendpoints/avatar.py", line 126, in avatar_post
django_1    |     response_dict = {'detail': rpc_handler(user, data)}
django_1    |   File "/app/misago/users/api/userendpoints/avatar.py", line 162, in avatar_gallery
django_1    |     avatars.gallery.set_avatar(user, image)
django_1    |   File "/app/misago/users/avatars/gallery.py", line 69, in set_avatar
django_1    |     store.store_new_avatar(user, Image.open(fp))
django_1    |   File "/usr/local/lib/python3.6/site-packages/PIL/Image.py", line 2586, in open
django_1    |     fp = io.BytesIO(fp.read())
django_1    | AttributeError: 'NoneType' object has no attribute 'read'
And I am curious, the code I provided above is already working, why bother with this problem.
```
def set_avatar(user, image):
    img = image.image.file
    store.store_new_avatar(user, Image.open(img))
```
Whether you use cloud storage or not, it works fine and there is no data to download.

**Comment by mr-ant** [comment]
Hi @rafalp
It's down,and work perfect.Thanks.

**Comment by rafalp** [comment]
Supersed by #1043

**Pull request closed without merging**